services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableQueryManagerImpl.java

```diff
@@ -285,13 +308,69 @@ QueryResultBundle queryAsStreamAfterAuthorization(ProgressCallback<Void> progres
 		Long count = null;
 		if(runCount){
 			// count requested.
-			count = runCountQuery(query, indexDao);
+			count = runCountQuery(queryToRun, indexDao);
 		}
+		
+		//run the facet counts if needed
+		if(returnFacets && queryFacetColumns != null && !queryFacetColumns.isEmpty()){
+			//use original query instead of queryToRun because need the where clause that was not modified by any facets
+			facetResults = runFacetQueries(query, queryFacetColumns, indexDao);
+		}
+		
+		//run 
+
 		bundle.setQueryResult(queryResult);
 		bundle.setQueryCount(count);
+		bundle.setFacets(facetResults);
 		return bundle;
 	}
-
+	
+	/**
+	 * Runs facet queries (enumeration count or range min/max) for all columns in queryFacetColumns. 
+	 * @param originalQuery the non-transformed query that was submitted by the user.
+	 * @param queryFacetColumns
+	 * @param indexDao
+	 * @return
+	 */
+	public static List<QueryFacetResultColumn> runFacetQueries(SqlQuery originalQuery,

```

Comment:
Create a FacetModel class that takes the Requested facets, and List<ColumnModel> (schema) in a constructor.
This will then have List<ValidatedQueryFacetColumn> as a member variable.  Also save any mapping used.
Add a method to generate the list of queries to be run (but not run the query).  The manager executes the queries and prepares the results.